File: AggUI/Font.cs
/*
This file is part of AggUI.

Copyright © 2003-2024, EPSITEC SA, 1400 Yverdon-les-Bains, Switzerland

AggUI is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
any later version.

AggUI is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

using System;
using System.IO;
using System.Reflection;
using System.Runtime.InteropServices;
using static AntigrainSharp.Native;

namespace AntigrainSharp
{
    public class Font : IDisposable
    {
        public static Font LoadFromFile(string fontname)
        {
            if (!File.Exists(fontname))
            {
                throw new FileNotFoundException($"No font file {fontname}");
            }
            IntPtr face = FreetypeInfo_LoadFromFile(Font.library, fontname);
            return new Font(face, fontname);
        }

        internal Font(IntPtr face, string fontname)
        {
            this.face = face;
            this.fontname = fontname;
        }

        ~Font(){
            this.Dispose();
        }

        public void Dispose(){
            if (this.face != IntPtr.Zero){
                FreetypeInfo_DoneFace(this.face);
                this.face = IntPtr.Zero;
            }
            this.fontname = "";
            GC.SuppressFinalize(this);
        }

        private void RequireNotDisposed(){
            if (this.face == IntPtr.Zero){
                throw new ObjectDisposedException(this.GetType().FullName);
            }
        }

        public bool IsBold
        {
            get {
                RequireNotDisposed();
                return FreetypeInfo_IsFaceBold(this.face);
            }
        }

        public bool IsItalic
        {
            get {
                RequireNotDisposed();
                return FreetypeInfo_IsFaceItalic(this.face);
            }
        }

        public uint UnitsPerEm
        {
            get {
                RequireNotDisposed();
                return FreetypeInfo_GetUnitsPerEm(this.face);
            }
        }

        public double GetGlyphAdvance(uint glyph, double size)
        {
            RequireNotDisposed();
            return FreetypeInfo_GetGlyphAdvance(this.face, glyph, size);
        }

        public uint GetCharIndex(ulong charcode)
        {
            RequireNotDisposed();
            return FreetypeInfo_GetCharIndex(this.face, charcode);
        }

        public void GetGlyphBBox(
            uint glyph,
            double size,
            out double xMin,
            out double xMax,
            out double yMin,
            out double yMax
        )
        {
            RequireNotDisposed();
            FreetypeInfo_GetGlyphBBox(
                this.face,
                glyph,
                size,
                out xMin,
                out xMax,
                out yMin,
                out yMax
            );
        }

        public double GetKerning(uint left_glyph, uint right_glyph, double size)
        {
            RequireNotDisposed();
            return FreetypeInfo_GetKerning(this.face, left_glyph, right_glyph, size);
        }

        public double GetAscender(double size)
        {
            RequireNotDisposed();
            return FreetypeInfo_GetAscender(this.face, size);
        }

        public double GetDescender(double size)
        {
            RequireNotDisposed();
            return FreetypeInfo_GetDescender(this.face, size);
        }

        public double GetHeight(double size)
        {
            RequireNotDisposed();
            return FreetypeInfo_GetHeight(this.face, size);
        }

        private IntPtr face;
        internal string fontname;

        private static IntPtr library = FreetypeInfo_CreateLibrary();
    }
}

File: AggUI/aggcpp/font.cpp
#include <ft2build.h>
#include <freetype/ftglyph.h>

#include "font.h"
#include "freetype_utils.h"


namespace AntigrainCPP {

    FT_Library FreetypeInfo_CreateLibrary(){
        FT_Library library;
        FT_Init_FreeType(&library);
        return library;
    }

    FT_Face FreetypeInfo_LoadFromFile(FT_Library library,
        const char* font_name
    ){
        FT_Face face;
        FT_New_Face(library, font_name, 0, &face); 
        SetFaceSize(face, 1.0);
        return face;
    }

    void FreetypeInfo_DoneFace(FT_Face face){
        FT_Done_Face(face);
    }

    bool FreetypeInfo_IsFaceBold(FT_Face face){
        return face->style_flags & FT_STYLE_FLAG_BOLD;
    }

    bool FreetypeInfo_IsFaceItalic(FT_Face face){
        return face->style_flags & FT_STYLE_FLAG_ITALIC;
    }

    unsigned FreetypeInfo_GetUnitsPerEm(FT_Face face){
        return (unsigned)face->units_per_EM;
    }

    unsigned FreetypeInfo_GetCharIndex(FT_Face face,
        unsigned long charcode
    ){
        return FT_Get_Char_Index(face, charcode);
    }

    double FreetypeInfo_GetGlyphAdvance(FT_Face face, unsigned glyph_index, double size){
        LoadGlyph(face, glyph_index, size);
        return F26Dot6ToDouble(face->glyph->metrics.horiAdvance);
    }

    double FreetypeInfo_GetKerning(FT_Face face, 
        unsigned left_glyph,
        unsigned right_glyph,
        double size
    ){
        SetFaceSize(face, size);
        FT_Vector kerning;
        FT_Get_Kerning(face, left_glyph, right_glyph, FT_KERNING_DEFAULT, &kerning);
        return F26Dot6ToDouble(kerning.x);
    }

    void LoadGlyph(FT_Face face, unsigned glyph_index, double size){
        SetFaceSize(face, size);
        FT_Load_Glyph(face, glyph_index, FT_LOAD_NO_BITMAP);
    }

    void SetFaceSize(FT_Face face, double size){
        FT_F26Dot6 height = doubleToF26Dot6(size);
        FT_Set_Char_Size(face, 0, height, 0, 0);
    }

    void FreetypeInfo_GetGlyphBBox(FT_Face face,
        unsigned glyph_index,
        double size,
        double &xMin,
        double &xMax,
        double &yMin,
        double &yMax
    ){
        LoadGlyph(face, glyph_index, size);
        FT_BBox bbox;
        FT_Glyph glyph;
        FT_Get_Glyph(face->glyph, &glyph);
        FT_Glyph_Get_CBox(glyph, FT_GLYPH_BBOX_GRIDFIT, &bbox);
        FT_Done_Glyph(glyph);
        xMin = F26Dot6ToDouble(bbox.xMin);
        xMax = F26Dot6ToDouble(bbox.xMax);
        yMin = F26Dot6ToDouble(bbox.yMin);
        yMax = F26Dot6ToDouble(bbox.yMax);
    }

    double FreetypeInfo_GetAscender(FT_Face face, double size)
    {
        SetFaceSize(face, size);
        return F26Dot6ToDouble(face->size->metrics.ascender);
    }

    double FreetypeInfo_GetDescender(FT_Face face, double size)
    {
        SetFaceSize(face, size);
        return F26Dot6ToDouble(face->size->metrics.descender);
    }

    double FreetypeInfo_GetHeight(FT_Face face, double size)
    {
        SetFaceSize(face, size);
        return F26Dot6ToDouble(face->size->metrics.height);
    }
}
